src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -6073,6 +6080,17 @@ int RGWRados::set_attrs(const DoutPrefixProvider *dpp, RGWObjectCtx* rctx, RGWBu
       r = index_op.complete(dpp, poolid, epoch, state->size, state->accounted_size,
                             mtime, etag, content_type, storage_class, &acl_bl,
                             RGWObjCategory::Main, nullptr, y);
+      if (r < 0)
+        return r;
+
+      if (versioned_op && epoch) {
+        constexpr bool log_data_change = true;
+        r = set_olh(dpp, *rctx, bucket_info, obj, false, nullptr, epoch, real_time(), false, y, nullptr, log_data_change);
```

Comment:
`epoch` and `olh_epoch` are two different things. i think you just want to pass `olh_epoch=0` into `set_olh()`, which should cause cls_rgw.cc's `rgw_bucket_link_olh()` -> `start_modify()` to increment its stored `olh_epoch`